ethdb/database.go

```diff
@@ -14,6 +14,8 @@
 // You should have received a copy of the GNU Lesser General Public License
 // along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
 
+// +build !js
+
```

Comment:
I like this idea, but please provide an API-compatible shim file for JS.

Reply:
`ethdb/database.go` imports `github.com/syndtr/goleveldb/leveldb` which is not compatible with the JS. There is `ethdb/memory_database.go` file with the `MemDatabase` type which implements all interfaces of `github.com/ethereum/go-ethereum/ethdb` package . So I did not quite understand, which API-compatible shim file should I provide. Could you please clarify?

Actually `ethdb/database.go` is excluded to make it possible to import `github.com/ethereum/go-ethereum/core/types` in WebAssembly application. Because `github.com/ethereum/go-ethereum/core/types` imports `github.com/ethereum/go-ethereum/trie`, and `github.com/ethereum/go-ethereum/trie` imports `github.com/ethereum/go-ethereum/ethdb`. I guess it's enough to have `ethdb.MemDatabase` type available in JS.

Reply:
What I meant is that there should be a file providing `NewLDBDatabase(...) (*LDBDatabase, error)` even for js. It can just return an error. So basically you would create a new file `database_js.go` which contains a dummy type with all the methods. The only method you can't add is `LDB() *leveldb.Database`. You can leave this one out for now.

An unrelated issue is that excluding database.go from the build also removes `NewTable`.
Please move `NewTable` and all the other table-related code to a separate file.

Reply:
Oh, I see. But there are actually 3 methods I can't add:

1. `LDB() *leveldb.Database`
2. `NewIterator() iterator.Iterator`
3. `NewIteratorWithPrefix(prefix []byte) iterator.Iterator`

2&3 - because they use `iterator.Iterator` interface from the `github.com/syndtr/goleveldb/leveldb/iterator` package

Please check my changes now.

Reply:
@fjl Running `go test` for `js/wasm` will fail now :( Is it ok?

```
$ GOOS=js GOARCH=wasm go test -exec="$(go env GOROOT)/misc/wasm/go_js_wasm_exec"

--- FAIL: TestLDB_PutGet (0.00s)
panic: failed to create test database: ethdb: not supported [recovered]
	panic: failed to create test database: ethdb: not supported

goroutine 5 [running]:
testing.tRunner.func1(0xc080100)
	/usr/local/go/src/testing/testing.go:792 +0x34
panic(0x1eec0, 0xc00e530)
	/usr/local/go/src/runtime/panic.go:513 +0x1d
github.com/ethereum/go-ethereum/ethdb_test.newTestLDB(0x14da0002, 0x5bb1f41a)
	/Users/dimchansky/Work/go/src/github.com/ethereum/go-ethereum/ethdb/database_test.go:38 +0x14
github.com/ethereum/go-ethereum/ethdb_test.TestLDB_PutGet(0xc080100)
	/Users/dimchansky/Work/go/src/github.com/ethereum/go-ethereum/ethdb/database_test.go:50 +0x2
testing.tRunner(0xc080100, 0x539e0)
	/usr/local/go/src/testing/testing.go:827 +0xc
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x24
exit status 2
FAIL	github.com/ethereum/go-ethereum/ethdb	1.081s
```

Reply:
@fjl Ok, the tests of `ethdb` package pass for all platforms now